File: facecat_java/src/facecat/topin/core/FCPaint.java
/*捂脸猫FaceCat框架 v1.0
 1.创始人-矿洞程序员-上海宁米科技创始人-脉脉KOL-陶德 (微信号:suade1984);
 2.联合创始人-上海宁米科技创始人-袁立涛(微信号:wx627378127);
 3.联合创始人-河北思尔企业管理咨询有限公司合伙人-肖添龙(微信号:xiaotianlong_luu);
 4.联合开发者-陈晓阳(微信号:chenxiaoyangzxy)，助理-朱炜(微信号:cnnic_zhu);
 5.该框架开源协议为BSD，欢迎对我们的创业活动进行各种支持，欢迎更多开发者加入。
 包含C/C++,Java,C#,iOS,MacOS,Linux六个版本的图形和通讯服务框架。
 */
package facecat.topin.core;

/**
 * 绘图类
 */
public interface FCPaint {

    /**
     * 添加曲线
     *
     * @param rect 矩形区域
     * @param startAngle 从 x 轴到弧线的起始点沿顺时针方向度量的角（以度为单位）
     * @param sweepAngle 从 startAngle 参数到弧线的结束点沿顺时针方向度量的角（以度为单位）
     */
    void addArc(FCRect rect, float startAngle, float sweepAngle);

    /**
     * 添加贝赛尔曲线
     *
     * @param point1 坐标1
     * @param point2 坐标2
     * @param point3 坐标3
     * @param point4 坐标4
     */
    void addBezier(FCPoint[] points);

    /**
     * 添加曲线
     *
     * @param points 点阵
     */
    void addCurve(FCPoint[] points);

    /**
     * 添加椭圆
     *
     * @param rect 矩形
     */
    void addEllipse(FCRect rect);

    /**
     * 添加直线
     *
     * @param x1 第一个点的横坐标
     * @param y1 第一个点的纵坐标
     * @param x2 第二个点的横坐标
     * @param y2 第二个点的纵坐标
     */
    void addLine(int x1, int y1, int x2, int y2);

    /**
     * 添加矩形
     *
     * @param rect 区域
     */
    void addRect(FCRect rect);

    /**
     * 添加扇形
     *
     * @param rect 矩形区域
     * @param startAngle 从 x 轴到弧线的起始点沿顺时针方向度量的角（以度为单位）
     * @param sweepAngle 从 startAngle 参数到弧线的结束点沿顺时针方向度量的角（以度为单位）
     */
    void addPie(FCRect rect, float startAngle, float sweepAngle);

    /**
     * 添加文字
     *
     * @param text 文字
     * @param font 字体
     * @param rect 区域
     */
    void addText(String text, FCFont font, FCRect rect);

    /**
     * 开始绘图
     *
     * @param hdc HDC
     * @param wRect 窗体区域
     * @param pRect 刷新区域
     */
    void beginPaint(int canvas, FCRect wRect, FCRect pRect);

    /**
     * 开始一段路径
     */
    void beginPath();

    /**
     * 清除缓存
     */
    void clearCaches();

    /**
     * 裁剪路径
     */
    void clipPath();

    /**
     * 闭合路径
     */
    void closeFigure();

    /**
     * 结束一段路径
     */
    void closePath();

    /**
     * 删除对象
     */
    void delete();

    /**
     * 绘制弧线
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param rect 矩形区域
     * @param startAngle 从 x 轴到弧线的起始点沿顺时针方向度量的角（以度为单位）
     * @param sweepAngle 从 startAngle 参数到弧线的结束点沿顺时针方向度量的角（以度为单位）
     */
    void drawArc(long dwPenColor, float width, int style, FCRect rect, float startAngle, float sweepAngle);

    /**
     * 设置贝赛尔曲线
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param points 坐标阵
     */
    void drawBezier(long dwPenColor, float width, int style, FCPoint[] points);

    /**
     * 绘制曲线
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param points 坐标阵
     */
    void drawCurve(long dwPenColor, float width, int style, FCPoint[] points);

    /**
     * 绘制矩形
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param rect 矩形区域
     */
    void drawEllipse(long dwPenColor, float width, int style, FCRect rect);

    /**
     * 绘制矩形
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param left 左侧坐标
     * @param top 顶部左标
     * @param right 右侧坐标
     * @param bottom 底部坐标
     */
    void drawEllipse(long dwPenColor, float width, int style, int left, int top, int right, int bottom);

    /**
     * 绘制图片
     *
     * @param imagePath 图片路径
     * @param rect 绘制区域
     */
    void drawImage(String imagePath, FCRect rect);

    /**
     * 绘制直线
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param x1 第一个点的横坐标
     * @param y1 第一个点的纵坐标
     * @param x2 第二个点的横坐标
     * @param y2 第二个点的纵坐标
     */
    void drawLine(long dwPenColor, float width, int style, int x1, int y1, int x2, int y2);

    /**
     * 绘制直线
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param x 第一个点坐标
     * @param y 第二个点的坐标
     */
    void drawLine(long dwPenColor, float width, int style, FCPoint x, FCPoint y);

    /**
     * 绘制路径
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     */
    void drawPath(long dwPenColor, float width, int style);

    /**
     * 绘制弧线
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param rect 矩形区域
     * @param startAngle 从 x 轴到弧线的起始点沿顺时针方向度量的角（以度为单位）
     * @param sweepAngle 从 startAngle 参数到弧线的结束点沿顺时针方向度量的角（以度为单位）
     */
    void drawPie(long dwPenColor, float width, int style, FCRect rect, float startAngle, float sweepAngle);

    /**
     * 绘制多边形
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param points 点的数组
     */
    void drawPolygon(long dwPenColor, float width, int style, FCPoint[] points);

    /**
     * 绘制大量直线
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param points 点集
     */
    void drawPolyline(long dwPenColor, float width, int style, FCPoint[] points);

    /**
     * 绘制矩形
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param rect 矩形区域
     */
    void drawRect(long dwPenColor, float width, int style, FCRect rect);

    /**
     * 绘制矩形
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param left 左侧坐标
     * @param top 顶部左标
     * @param right 右侧坐标
     * @param bottom 底部坐标
     */
    void drawRect(long dwPenColor, float width, int style, int left, int top, int right, int bottom);

    /**
     * 绘制圆角矩形
     *
     * @param dwPenColor 颜色
     * @param width 宽度
     * @param style 样式
     * @param rect 矩形区域
     * @param cornerRadius 边角半径
     */
    void drawRoundRect(long dwPenColor, float width, int style, FCRect rect, int cornerRadius);

    /**
     * 绘制矩形
     *
     * @param text 文字
     * @param dwPenColor 颜色
     * @param font 字体
     * @param rect 矩形区域
     */
    void drawText(String text, long dwPenColor, FCFont font, FCRect rect);

    /**
     * 绘制矩形
     *
     * @param text 文字
     * @param dwPenColor 颜色
     * @param font 字体
     * @param rect 矩形区域
     */
    void drawText(String text, long dwPenColor, FCFont font, FCRectF rect);

    /**
     * 绘制自动省略结尾的文字
     *
     * @param text 文字
     * @param dwPenColor 颜色
     * @param font 字体
     * @param rect 矩形区域
     */
    void drawTextAutoEllipsis(String text, long dwPenColor, FCFont font, FCRect rect);

    /**
     * 结束导出
     */
    void endPaint();

    /**
     * 结束绘图
     */
    void excludeClipPath();

    /**
     * 填充椭圆
     *
     * @param dwPenColor 颜色
     * @param rect 矩形区域
     */
    void fillEllipse(long dwPenColor, FCRect rect);

    /**
     * 填充椭圆
     *
     * @param dwPenColor 颜色
     * @param left 左侧坐标
     * @param top 顶部左标
     * @param right 右侧坐标
     * @param bottom 底部坐标
     */
    void fillEllipse(long dwPenColor, int left, int top, int right, int bottom);

    /**
     * 绘制渐变椭圆
     *
     * @param dwFirst 开始颜色
     * @param dwSecond 结束颜色
     * @param rect 矩形
     * @param angle 角度
     */
    void fillGradientEllipse(long dwFirst, long dwSecond, FCRect rect, int angle);

    /**
     * 填充渐变路径
     *
     * @param dwFirst 开始颜色
     * @param dwSecond 结束颜色
     * @param points 点的集合
     * @param angle 角度
     */
    void fillGradientPath(long dwFirst, long dwSecond, FCRect rect, int angle);

    /**
     * 绘制渐变的多边形
     *
     * @param dwFirst 开始颜色
     * @param dwSecond 结束颜色
     * @param points 点的集合
     * @param angle 角度
     */
    void fillGradientPolygon(long dwFirst, long dwSecond, FCPoint[] points, int angle);

    /**
     * 绘制渐变矩形
     *
     * @param dwFirst 开始颜色
     * @param dwSecond 结束颜色
     * @param rect 矩形
     * @param cornerRadius 圆角半径
     * @param angle 角度
     */
    void fillGradientRect(long dwFirst, long dwSecond, FCRect rect, int cornerRadius, int angle);

    /**
     * 填充路径
     *
     * @param dwPenColor 颜色
     */
    void fillPath(long dwPenColor);

    /**
     * 绘制扇形
     *
     * @param dwPenColor 颜色
     * @param rect 矩形区域
     * @param startAngle 从 x 轴到弧线的起始点沿顺时针方向度量的角（以度为单位）
     * @param sweepAngle 从 startAngle 参数到弧线的结束点沿顺时针方向度量的角（以度为单位）
     */
    void fillPie(long dwPenColor, FCRect rect, float startAngle, float sweepAngle);

    /**
     * 填充椭圆
     *
     * @param dwPenColor 颜色
     * @param points 点的数组
     */
    void fillPolygon(long dwPenColor, FCPoint[] points);

    /**
     * 填充矩形
     *
     * @param dwPenColor 颜色
     * @param rect 矩形区域
     */
    void fillRect(long dwPenColor, FCRect rect);

    /**
     * 填充矩形
     *
     * @param text 文字
     * @param dwPenColor 颜色
     * @param font 字体
     * @param rect 矩形区域
     */
    void fillRect(long dwPenColor, int left, int top, int right, int bottom);

    /**
     * 填充圆角矩形
     *
     * @param dwPenColor 颜色
     * @param rect 矩形区域
     * @param cornerRadius 边角半径
     */
    void fillRoundRect(long dwPenColor, FCRect rect, int cornerRadius);

    /**
     * 获取颜色
     *
     * @param dwPenColor 输入颜色
     * @returns 输出颜色
     */
    long getColor(long dwPenColor);

    /**
     * 获取要绘制的颜色
     *
     * @param dwPenColor 输入颜色
     * @returns 输出颜色
     */
    long getPaintColor(long dwPenColor);

    /**
     * 获取偏移
     */
    FCPoint getOffset();

    /**
     * 旋转角度
     *
     * @param op 圆心坐标
     * @param mp 点的坐标
     * @param angle 角度
     * @returns 结果坐标
     */
    FCPoint rotate(FCPoint op, FCPoint mp, int angle);

    /**
     * 设置裁剪区域
     *
     * @param rect 区域
     */
    void setClip(FCRect rect);

    /**
     * 设置直线两端的样式
     *
     * @param startLineCap 开始的样式
     * @param endLineCap 结束的样式
     */
    void setLineCap(int startLineCap, int endLineCap);

    /**
     * 设置偏移
     *
     * @param mp 偏移坐标
     */
    void setOffset(FCPoint mp);

    /**
     * 设置透明度
     *
     * @param opacity 透明度
     */
    void setOpacity(float opacity);

    /**
     * 设置资源的路径
     *
     * @param resourcePath 资源的路径
     */
    void setResourcePath(String resourcePath);

    /**
     * 设置旋转角度
     *
     * @param rotateAngle 旋转角度
     */
    void setRotateAngle(int rotateAngle);

    /**
     * 设置缩放因子
     *
     * @param scaleFactorX 横向因子
     * @param scaleFactorY 纵向因子
     */
    void setScaleFactor(double scaleFactorX, double scaleFactorY);

    /**
     * 设置平滑模式
     *
     * @param smoothMode 平滑模式
     */
    void setSmoothMode(int smoothMode);

    /**
     * 设置文字的质量
     *
     * @param textQuality 文字质量
     */
    void setTextQuality(int textQuality);

    /**
     * 设置是否支持透明色
     *
     * @returns 是否支持
     */
    boolean supportTransparent();

    /**
     * 获取文字大小
     *
     * @param text 文字
     * @param font 字体
     * @returns 字体大小
     */
    FCSize textSize(String text, FCFont font);

    /**
     * 获取文字大小
     *
     * @param text 文字
     * @param font 字体
     * @returns 字体大小
     */
    FCSizeF textSizeF(String text, FCFont font);
}

File: facecat_cpp/source/core/FCPaint.cpp
#include "..\\..\\stdafx.h"
#include "..\\..\\include\\core\\FCPaint.h"

namespace FaceCat{
	Long FCColor::argb(int r, int g, int b){
		return ((r | (g << 8)) | (b << 0x10));

	}

	Long FCColor::argb(int a, int r, int g, int b){
		if(a == 255){
			return ((r | (g << 8)) | (b << 0x10));
		}
		else if(a == 0){
			return FCColor_None;
		}
		else{
			int rgb = ((r | (g << 8)) | (b << 0x10));
			Long argb = -((Long)rgb * 1000 + a);
			return argb;
		}
	}

	void FCColor::toArgb(FCPaint *paint, Long dwPenColor, int *a, int *r, int *g, int *b){
		if(paint){
			dwPenColor = paint->getColor(dwPenColor);
		}
		*a = 255;
		if(dwPenColor < 0){
			dwPenColor = -dwPenColor;
			if(dwPenColor < 1){
				*a = 255;
			}
			else{
				*a = (int)(dwPenColor - dwPenColor / 1000 * 1000);
			}
			dwPenColor /= 1000;
		}
		*r = (int)(dwPenColor & 0xff);
		*g = (int)((dwPenColor >> 8) & 0xff);
		*b = (int)((dwPenColor >> 0x10) & 0xff);
	}

	Long FCColor::ratioColor(FCPaint *paint, Long originalColor, double ratio){
		int a = 0, r = 0, g = 0, b = 0;
        toArgb(paint, originalColor, &a, &r, &g, &b);
        r = (int)(r * ratio);
        g = (int)(g * ratio);
        b = (int)(b * ratio);
        if (r > 255) r = 255;
        if (g > 255) g = 255;
        if (b > 255) b = 255;
        return argb(a, r, g, b);
	}

	Long FCColor::reverse(FCPaint *paint, Long originalColor){
		int a = 0, r = 0, g = 0, b = 0;
		toArgb(paint, originalColor, &a, &r, &g, &b);
		return argb(a, 255 - r, 255 - g, 255 - b);
	}

	///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	FCPaint::FCPaint(){
	}

	FCPaint::~FCPaint(){
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	void FCPaint::addArc(const FCRect& rect, float startAngle, float sweepAngle){
	}
	
	void FCPaint::addBezier(FCPoint *apt, int cpt){
	}
	
	void FCPaint::addCurve(FCPoint *apt, int cpt){
	}
	
	void FCPaint::addEllipse(const FCRect& rect){
	}
	
	void FCPaint::addLine(int x1, int y1, int x2, int y2){
	}
	
	void FCPaint::addRect(const FCRect& rect){
	}
	
	void FCPaint::addPie(const FCRect& rect, float startAngle, float sweepAngle){
	}
	
	void FCPaint::addText(const wchar_t *strText, FCFont *font, const FCRect& rect){
	}

	void FCPaint::beginExport(const String& exportPath, const FCRect& rect){
	}
    
	void FCPaint::beginPaint(HDC hDC, const FCRect& wRect, const FCRect& pRect){
        
	}
	
	void FCPaint::beginPath(){
	}
    
	void FCPaint::clearCaches(){
	}

	void FCPaint::clipPath(){
	}
	
	void FCPaint::closeFigure(){
	}
	
	void FCPaint::closePath(){
	}
	
	void FCPaint::drawArc(Long dwPenColor, float width, int style, const FCRect& rect, float startAngle, float sweepAngle){
	}
	
	void FCPaint::drawBezier(Long dwPenColor, float width, int style, FCPoint *apt, int cpt){
	}
	
	void FCPaint::drawCurve(Long dwPenColor, float width, int style, FCPoint *apt, int cpt){
	}
    
	void FCPaint::drawEllipse(Long dwPenColor, float width, int style, const FCRect& rect){
	}
    
	void FCPaint::drawEllipse(Long dwPenColor, float width, int style, int left, int top, int right, int bottom){
	}
    
	void FCPaint::drawImage(const wchar_t *imagePath, const FCRect& rect){
	}
    
	void FCPaint::drawLine(Long dwPenColor, float width, int style, const FCPoint& x, const FCPoint& y){
	}
    
	void FCPaint::drawLine(Long dwPenColor, float width, int style, int x1, int y1, int x2, int y2){
	}
	
	void FCPaint::drawPath(Long dwPenColor, float width, int style){
	}
	
	void FCPaint::drawPie(Long dwPenColor, float width, int style, const FCRect& rect, float startAngle, float sweepAngle){
	}
    
	void FCPaint::drawPolygon(Long dwPenColor, float width, int style, FCPoint *apt, int cpt){
	}
    
	void FCPaint::drawPolyline(Long dwPenColor, float width, int style, FCPoint *apt, int cpt){
	}
    
	void FCPaint::drawRect(Long dwPenColor, float width, int style, int left, int top, int right, int bottom){
	}
    
	void FCPaint::drawRect(Long dwPenColor, float width, int style, const FCRect& rect){
	}
    
	void FCPaint::drawRoundRect(Long dwPenColor, float width, int style, const FCRect& rect, int cornerRadius){
	}
    
	void FCPaint::drawText(const wchar_t *strText, Long dwPenColor, FCFont *font, const FCRect& rect){
	}
    
	void FCPaint::drawText(const wchar_t *strText, Long dwPenColor, FCFont *font, const FCRectF& rect){
	}
    
	void FCPaint::drawTextAutoEllipsis(const wchar_t *strText, Long dwPenColor, FCFont *font, const FCRect& rect){
	}

	void FCPaint::endExport(){
	}
    
	void FCPaint::endPaint(){
	}

	void FCPaint::excludeClipPath(){
	}
    
	void FCPaint::fillEllipse(Long dwPenColor, const FCRect& rect){
	}
	
	void FCPaint::fillGradientEllipse(Long dwFirst, Long dwSecond, const FCRect& rect, int angle){
	}

	void FCPaint::fillGradientPath(Long dwFirst, Long dwSecond, const FCRect& rect, int angle){
	}

	void FCPaint::fillGradientPolygon(Long dwFirst, Long dwSecond, FCPoint *apt, int cpt, int angle){
	}
    
	void FCPaint::fillGradientRect(Long dwFirst, Long dwSecond, const FCRect& rect, int cornerRadius, int angle){
	}
	
	void FCPaint::fillPath(Long dwPenColor){
	}
	
	void FCPaint::fillPie(Long dwPenColor, const FCRect& rect, float startAngle, float sweepAngle){
	}
    
	void FCPaint::fillPolygon(Long dwPenColor, FCPoint *apt, int cpt){
	}
    
	void FCPaint::fillRect(Long dwPenColor, const FCRect& rect){
	}
    
	void FCPaint::fillRect(Long dwPenColor, int left, int top, int right, int bottom){
	}
    
	void FCPaint::fillRoundRect(Long dwPenColor, const FCRect& rect, int cornerRadius){
	}
    
	Long FCPaint::getColor(Long dwPenColor){
		return 0;
	}
    
	Long FCPaint::getPaintColor(Long dwPenColor){
		return 0;
	}
    
	FCPoint FCPaint::getOffset(){
		FCPoint offset = {0};
		return offset;
	}
    
    FCPoint FCPaint::rotate(const FCPoint& op, const FCPoint& mp, int angle){
		FCPoint pt = {0};
		return pt;
	}
    
	void FCPaint::setClip(const FCRect& rect){
	}
	
	void FCPaint::setLineCap(int startLineCap, int endLineCap){
	}
    
	void FCPaint::setOffset(const FCPoint& offset){
        
	}
    
	void FCPaint::setOpacity(float opacity){
	}
    
	void FCPaint::setResourcePath(const String& resourcePath){
	}
    
    void FCPaint::setRotateAngle(int rotateAngle){
        
    }
    
	void FCPaint::setScaleFactor(double scaleFactorX, double scaleFactorY){
	}

	void FCPaint::setSmoothMode(int smoothMode){
	}

	void FCPaint::setTextQuality(int textQuality){
	}
    
	bool FCPaint::supportTransparent(){
		return false;
	}
    
	FCSize FCPaint::textSize(const wchar_t *strText, FCFont *font){
		FCSize size = {0};
		return size;
	}
    
	FCSizeF FCPaint::textSizeF(const wchar_t *strText, FCFont *font){
		FCSizeF size = {0};
		return size;
	}
}